<a href="https://colab.research.google.com/github/marce3-2140/Data-Analysis-with-Panda/blob/main/AlturaPromedioPorOlimpico.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>



# **DataFrame name: 120 Years of Olimpyc History**



> **Contexto:**

This is a historical dataset on the modern Olympic Games, Including all the games from Athens 1896 to Rio 2016.

>  **Descripción del Diccionario de Datos**

[ByTableGenerator](https://www.tablesgenerator.com/markdown_tables#)

>**Desrcipción de la actividad**

1.  Carguen este conjunto de datos y explórenlo. Respondan las siguientes preguntas:

     a. ¿Cuántas filas hay?

     b. ¿Cuántas columnas hay?

     c. ¿A qué columnas le faltan datos? (¿Por qué creen que hay muchos valores nulos en la columna Medals (medallas)?

     d. ¿Cuántas entradas corresponden a la ciudad de Londres?

     e. ¿Cuántos años tiene el atleta más joven en nuestros datos de muestra?  Pista: usen min()

2. Creen filtros para encontrar:

     a. ¿Cuántos atletas que participaron en el deporte "Tug-Of-War" (tira y afloja) pertenecían en el equipo “Sweden” (sueco) en el conjunto de datos de muestra?

     b. ¿Cuántas medallas de oro se le entregaron a los atletas que tenían más de 40 años?

3. Determinen la altura promedio de:

     a. Todas las filas incluidas.

     b. Todas las filas incluidas, pero agrupadas por Event (evento).

     c. Bono: Este conjunto de datos contiene muchos atletas más de una vez (si compitieron en varios eventos y/o varios años). Pueden confirmarlo encontrarlo .value_counts en la columna ID.  Para la tarea del bono queremos que obtengan la altura promedio de todos los atletas, pero sin incluir algún atleta más de una vez.  Pista: Aprendimos cómo eliminar duplicados si CADA columna fuera idéntica, sin embargo, ahora tenemos que eliminar una fila cada vez que un ID se repite.  Para hacer eso, pueden usar un subconjunto de sus datos como un argumento en la función drop_duplicates.  También le pueden decir a Python cuales duplicados mantener usando el argumento keep.

    

```
Because I don't want to drop all of these values permanently from my df,
assign this as a new dataframe that I called "one_per_athlete"
one_per_athlete = df.drop_duplicates(subset = 'Column_Name', keep = 'first')


```



### **Solución:**

In [48]:
# librerias

import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns

In [49]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [50]:
filename = '/content/drive/MyDrive/CodingDojo/Fundamentals/Week3/Cargas/athleteEventsNoPersonal.csv'
df = pd.read_csv(filename)

In [51]:
from google.colab import drive
drive.mount('/content/drive')

filename = '/content/drive/MyDrive/CodingDojo/Fundamentals/Week3/Cargas/athleteEventsNoPersonal.csv'
df = pd.read_csv(filename)

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


###I. **Análisis Exploratorio de los Datos**

In [52]:
df.head()

ID   Age  Height         Team  NOC        Games  Year  Season         City  \
0   5  21.0   185.0  Netherlands  NED  1988 Winter  1988  Winter      Calgary   
1   5  21.0   185.0  Netherlands  NED  1988 Winter  1988  Winter      Calgary   
2   5  25.0   185.0  Netherlands  NED  1992 Winter  1992  Winter  Albertville   
3   5  25.0   185.0  Netherlands  NED  1992 Winter  1992  Winter  Albertville   
4   5  27.0   185.0  Netherlands  NED  1994 Winter  1994  Winter  Lillehammer   

           Sport                               Event Medal  
0  Speed Skating    Speed Skating Women's 500 metres   NaN  
1  Speed Skating  Speed Skating Women's 1,000 metres   NaN  
2  Speed Skating    Speed Skating Women's 500 metres   NaN  
3  Speed Skating  Speed Skating Women's 1,000 metres   NaN  
4  Speed Skating    Speed Skating Women's 500 metres   NaN

In [53]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40616 entries, 0 to 40615
Data columns (total 12 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   ID      40616 non-null  int64  
 1   Age     39143 non-null  float64
 2   Height  31615 non-null  float64
 3   Team    40616 non-null  object 
 4   NOC     40616 non-null  object 
 5   Games   40616 non-null  object 
 6   Year    40616 non-null  int64  
 7   Season  40616 non-null  object 
 8   City    40616 non-null  object 
 9   Sport   40616 non-null  object 
 10  Event   40616 non-null  object 
 11  Medal   5917 non-null   object 
dtypes: float64(2), int64(2), object(8)
memory usage: 3.7+ MB


**1. Carguen este conjunto de datos y explórenlo. Respondan las siguientes preguntas:**

In [54]:
#a. ¿Cuántas filas hay?
#b. ¿Cuántas columnas hay?
df.shape

(40616, 12)

In [55]:
#c. ¿A qué columnas le faltan datos? (¿Por qué creen que hay muchos valores nulos en la columna Medals (medallas)?
df.isnull().sum()

ID            0
Age        1473
Height     9001
Team          0
NOC           0
Games         0
Year          0
Season        0
City          0
Sport         0
Event         0
Medal     34699
dtype: int64

In [56]:
#Filtramos los datos nulos de la columna Medals para identificar con que otra columna puede relacionarse
medals_missing = df['Medal'].isnull() #filtro para valores NaN
df.loc[medals_missing,:].head(10)

ID   Age  Height           Team  NOC        Games  Year  Season  \
0   5  21.0   185.0    Netherlands  NED  1988 Winter  1988  Winter   
1   5  21.0   185.0    Netherlands  NED  1988 Winter  1988  Winter   
2   5  25.0   185.0    Netherlands  NED  1992 Winter  1992  Winter   
3   5  25.0   185.0    Netherlands  NED  1992 Winter  1992  Winter   
4   5  27.0   185.0    Netherlands  NED  1994 Winter  1994  Winter   
5   5  27.0   185.0    Netherlands  NED  1994 Winter  1994  Winter   
6   6  31.0   188.0  United States  USA  1992 Winter  1992  Winter   
7   6  31.0   188.0  United States  USA  1992 Winter  1992  Winter   
8   6  31.0   188.0  United States  USA  1992 Winter  1992  Winter   
9   6  31.0   188.0  United States  USA  1992 Winter  1992  Winter   

          City                 Sport  \
0      Calgary         Speed Skating   
1      Calgary         Speed Skating   
2  Albertville         Speed Skating   
3  Albertville         Speed Skating   
4  Lillehammer         Speed Skating   
5  Lillehammer         Speed Skating   
6  Albertville  Cross Country Skiing   
7  Albertville  Cross Country Skiing   
8  Albertville  Cross Country Skiing   
9  Albertville  Cross Country Skiing   

                                               Event Medal  
0                   Speed Skating Women's 500 metres   NaN  
1                 Speed Skating Women's 1,000 metres   NaN  
2                   Speed Skating Women's 500 metres   NaN  
3                 Speed Skating Women's 1,000 metres   NaN  
4                   Speed Skating Women's 500 metres   NaN  
5                 Speed Skating Women's 1,000 metres   NaN  
6           Cross Country Skiing Men's 10 kilometres   NaN  
7           Cross Country Skiing Men's 50 kilometres   NaN  
8  Cross Country Skiing Men's 10/15 kilometres Pu...   NaN  
9  Cross Country Skiing Men's 4 x 10 kilometres R...   NaN

Al parecer una de las razones por las que faltan muchos datos en la columna Medals que se observa al ver la tabla anterior es que hay muchas filas duplicadas, revisaremos cuantos duplicados hay

In [57]:
df.duplicated().sum()

207

In [58]:
#Cuantos son los datos faltamtes en la Columna Medals
sum_medal_missing = df['Medal'].isna().sum()
total_medal_missing = (sum_medal_missing/len(df))*100
print(f"La columna 'Medal' tiene {total_medal_missing:.2f}% de datos nulos")

La columna 'Medal' tiene 85.43% de datos nulos


In [59]:
#Age
#Cuantos son los datos faltamtes en la Columna Age
sum_age_missing = df['Age'].isna().sum()
total_age_missing = (sum_age_missing/len(df))*100
print(f"La columna 'Age' tiene {total_age_missing:.2f}% de datos nulos")

La columna 'Age' tiene 3.63% de datos nulos


In [60]:
#Height
#Cuantos son los datos faltamtes en la Columna Height
sum_Height_missing = df['Height'].isna().sum()
total_Height_missing = (sum_Height_missing/len(df))*100
print(f"La columna 'Height' tiene {total_Height_missing:.2f}% de datos nulos")

La columna 'Height' tiene 22.16% de datos nulos


In [61]:
#d. ¿Cuántas entradas corresponden a la ciudad de Londres?
df_filtrado = df['City']=='London'
df_filtrado.value_counts()

False    37246
True      3370
Name: City, dtype: int64

In [62]:
#e. ¿Cuántos años tiene el atleta más joven en nuestros datos de muestra? Pista: usen min()
df['Age'].min()

11.0

El atleta más joven en los datos tiene 11 años.

**2. Creen filtros para encontrar:**

In [63]:
#¿Cuántos atletas que participaron en el deporte "Tug-Of-War" (tira y afloja) pertenecían en el equipo “Sweden” (sueco) en el conjunto de datos de muestra?
df_sports = df['Sport']=='Tug-Of-War'
df_team = df['Team']=='Sweden'
df_sports_team = (df_sports & df_team)
df_sports_team.value_counts()

False    40610
True         6
dtype: int64

6 atletas que participaron en el deporte "Tug-Of-War" (tira y afloja) pertenecían en el equipo “Sweden” (sueco) en el conjunto de datos de muestra.

In [64]:
#¿Cuántas medallas de oro se le entregaron a los atletas que tenían más de 40 años?
df_40 = df['Age']>40
df_medals_oro = df['Medal']=='Gold'
df_40_oro = (df_40 & df_medals_oro)
df_40_oro.sum()

66

In [65]:
df['Medal'].value_counts()

Gold      2009
Bronze    1994
Silver    1914
Name: Medal, dtype: int64

66 medallas de oro de un total de 2009, se le entregaron a los atletas que tenían más de 40 años.

**3. Determinen la altura promedio**


In [68]:
#a. Todas las filas incluidas.
df['Height'].mean().round(1)

175.4

La altura promedio con todas las filas incluidas es de 175 cms

In [69]:
#b. Todas las filas incluidas, pero agrupadas por Event (evento).
df_evento = df.groupby('Event')['Height'].mean().round(1)
df_evento

Event
Alpine Skiing Men's Combined                      178.5
Alpine Skiing Men's Downhill                      178.2
Alpine Skiing Men's Giant Slalom                  176.7
Alpine Skiing Men's Slalom                        176.5
Alpine Skiing Men's Super G                       179.9
                                                  ...  
Wrestling Women's Flyweight, Freestyle            158.1
Wrestling Women's Heavyweight, Freestyle          173.4
Wrestling Women's Light-Heavyweight, Freestyle    170.0
Wrestling Women's Lightweight, Freestyle          162.7
Wrestling Women's Middleweight, Freestyle         164.5
Name: Height, Length: 726, dtype: float64

**c. Bono**: Este conjunto de datos contiene muchos atletas más de una vez (si compitieron en varios eventos y/o varios años). Pueden confirmarlo encontrarlo .value_counts en la columna ID. Para la tarea del bono queremos que obtengan la altura promedio de todos los atletas, pero sin incluir algún atleta más de una vez. Pista: Aprendimos cómo eliminar duplicados si CADA columna fuera idéntica, sin embargo, ahora tenemos que eliminar una fila cada vez que un ID se repite. Para hacer eso, pueden usar un subconjunto de sus datos como un argumento en la función drop_duplicates. También le pueden decir a Python cuales duplicados mantener usando el argumento keep.

In [70]:
df['ID'].value_counts()

119591    36
44875     32
76437     31
133279    30
21462     29
          ..
78467      1
78468      1
78470      1
78474      1
91368      1
Name: ID, Length: 20336, dtype: int64

In [72]:
#Because I don't want to drop all of these values permanently from my df,
#assign this as a new dataframe that I called "one_per_athlete"
one_per_athlete = df.drop_duplicates(subset = 'Height', keep = 'first')

In [73]:
one_per_athlete['ID'].value_counts()

5         1
9407      1
7769      1
7283      1
7109      1
         ..
526       1
452       1
351       1
336       1
111572    1
Name: ID, Length: 82, dtype: int64

In [75]:
one_per_athlete['Height'].mean().round(1)

178.0

La Altura promedio de los atletas sin aparecer más de una vez es de 178 cms.

###V. **Referencias**


* [Curso Coding Dojo - Data Science](https://github.com/fralfaro/CD-DS-Course)
* [Kaggle - Competitions](https://www.kaggle.com/competitions)
* [fralfaro-github-vusualization](https://fralfaro.github.io/MAT281_2022/lectures/data_manipulation/visualization/seaborn/)